mathesar_ui/src/component-library/debounce/Debounce.svelte

```diff
@@ -0,0 +1,61 @@
+<script lang="ts">
+  import { createEventDispatcher } from 'svelte';
+  import type { ArtificialEvents } from '@mathesar-component-library-dir/common/types/ArtificialEvents';
+
+  const dispatch = createEventDispatcher<ArtificialEvents<unknown>>();
+
+  export let duration = 300;
+  let parentValue: unknown = undefined;
+  export { parentValue as value };
+
+  let timeout: number;
+
+  function getValueFromArtificialEvent(event: CustomEvent<unknown>) {
+    return event.detail;
+  }
+
+  function getValueFromStandardEvent(e: InputEvent): unknown {
+    if (!e.target) {
+      return undefined;
+    }
+    const target = e.target as { value?: unknown };
+    return target.value;
+  }
+
+  function setParentValue(v: unknown) {
+    parentValue = v;
+    dispatch('artificialInput', v);
+    dispatch('artificialChange', v);
+  }
+
+  function handleGenericInput(v: unknown) {
+    clearTimeout(timeout);
+    timeout = window.setTimeout(() => {
+      setParentValue(v);
+    }, duration);
+  }
+  function handleArtificialInput(e: CustomEvent<unknown>) {
+    handleGenericInput(getValueFromArtificialEvent(e));
+  }
+  function handleStandardInput(e: InputEvent) {
+    handleGenericInput(getValueFromStandardEvent(e));
+  }
+
+  function handleGenericChange(v: unknown) {
+    clearTimeout(timeout);
+    setParentValue(v);
+  }
+  function handleArtificialChange(e: CustomEvent<unknown>) {
+    handleGenericChange(getValueFromArtificialEvent(e));
+  }
+  function handleStandardChange(e: InputEvent) {
+    handleGenericChange(getValueFromStandardEvent(e));
+  }
+</script>
+
+<slot
+  artificialInput={handleArtificialInput}
+  input={handleStandardInput}
+  artificialChange={handleArtificialChange}
+  change={handleStandardChange}
+/>
```

Comment:
Thanks for this suggestion. I didn't like this design in my PR either. I've added 26091ce179f2affc9256a0877fc1b21a65528d7d to improve it along the lines that you've suggested. The `Debounce` component is now decoupled from the concerns of `input` vs `artificialInput`. I wasn't able to make the `update` function quite as simple as your example, because the `Debounce` component needs to handle events differently depending on whether they should be debounced. I hope the design I went with will suffice

Reply:
Looks good to me